src/main/scala/tech/cryptonomic/conseil/Lorre.scala

```diff
@@ -1,771 +1,53 @@
-    Try {
-      Await.result(tezosNodeOperator.getBareBlockHead(), lorreConf.bootupConnectionCheckTimeout)
-    } match {
-      case Failure(e) =>
-        logger.error("Could not make initial connection to Tezos", e)
-        Thread.sleep(lorreConf.bootupRetryInterval.toMillis)
-        checkTezosConnection()
-      case Success(_) =>
-        logger.info("Successfully made initial connection to Tezos")
-    }
-
-  // Finds unprocessed levels for account refreshes (i.e. when there is a need to reload all accounts data from the chain)
-  private def unprocessedLevelsForRefreshingAccounts(): Future[AccountUpdatesEvents] =
-    lorreConf.chainEvents.collectFirst {
-      case ChainEvent.AccountsRefresh(levelsNeedingRefresh) if levelsNeedingRefresh.nonEmpty =>
-        db.run(TezosDb.fetchProcessedEventsLevels(ChainEvent.accountsRefresh.render)).map { levels =>
-          //used to remove processed events
-          val processed = levels.map(_.intValue).toSet
-          //we want individual event levels with the associated pattern, such that we can sort them by level
-          val unprocessedEvents = levelsNeedingRefresh.toList.flatMap {
-            case (accountPattern, levels) => levels.filterNot(processed).sorted.map(_ -> accountPattern)
-          }
-          SortedSet(unprocessedEvents: _*)
-        }
-    }.getOrElse(Future.successful(SortedSet.empty))
-
-  /** The regular loop, once connection with the node is established */
-  @tailrec
-  private[this] def mainLoop(
-      iteration: Int,
-      accountsRefreshLevels: AccountUpdatesEvents
-  ): Unit = {
-    val noOp = Future.successful(())
-    val processing = for {
-      nextRefreshes <- processAccountRefreshes(accountsRefreshLevels)
-      _ <- processTezosBlocks()
-      _ <- if (iteration % lorreConf.feeUpdateInterval == 0)
-        FeeOperations.processTezosAverageFees(lorreConf.numberOfFeesAveraged)
-      else
-        noOp
-    } yield Some(nextRefreshes)
-
-    /* Won't stop Lorre on failure from processing the chain, unless overridden by the environment to halt.
-     * Can be used to investigate issues on consistently failing block or account processing.
-     * Otherwise, any error will make Lorre proceed as expected by default (stop or wait for next cycle)
+object Lorre extends IOApp with LazyLogging with IOLogging with LorreAppConfig {
+
+  //defines the main logic, given all necessary dependencies and configurations
+  override def run(args: List[String]): IO[ExitCode] = {
+    import LorreAppConfig.ConfiguredResources
+    import cats.arrow.Arrow
+    import cats.instances.function._
+
+    //builds operations on which Lorre depends, applying the configured resources
+    //it uses cats.arrow idioms to pair the input with the output
+    val buildDeps: ConfiguredResources => (ConfiguredResources, application.CompositeOps) =
+      Arrow[Function].lift(identity[ConfiguredResources]).merge((application.makeOperations _).tupled)
```

Comment:
rewrite this simply as something like: 
```
res => {
  val ops = makeOps(res)
  (res, ops)
}
```

Reply:
The idea here is that we need a function that builds on top of "general" resources (database connection pool, actor system, execution context, purely functional async context, ...) a set (tuple) of "specialized" operators/operation-set/services.
ConfiguredResources contains the general resources, built from configuration params.
application.CompositeOps is the set of specialized operations.
They're created from the ConfiguredResources using the `application.makeOperations` method.

The `Arrow` plumbing is only needed to make the types fit correctly, based on the assumption that `makeOperations` is to be written in the most straightforward way, without any hint to _how_ it will be used.

Here we make use of `cats.arrow.Arrow` to be able to convert the makeOperations function such that it also returns the passed-in argument as part of the output.

We'll see that this is needed later on so that both will be available later in the actual code that runs the application, which needs both general and specialized resources.